Sources_v3/Database/DTOs/ChannelReadDTO.swift

```diff
@@ -0,0 +1,66 @@
+//
+// Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import CoreData
+import Foundation
+
+@objc(ChannelReadDTO)
+class ChannelReadDTO: NSManagedObject {
+    static let entityName = "ChannelReadDTO"
+    
+    @NSManaged var lastReadAt: Date
+    @NSManaged var unreadMessageCount: Int
+    
+    // MARK: - Relationships
+    
+    @NSManaged var channel: ChannelDTO
+    @NSManaged var user: UserDTO
+    
+    static func fetchRequest(for cid: ChannelId, userId: String) -> NSFetchRequest<ChannelReadDTO> {
+        let request = NSFetchRequest<ChannelReadDTO>(entityName: ChannelReadDTO.entityName)
+        request.predicate = NSPredicate(format: "channel.cid == %@ && user.id == %@", cid.rawValue, userId)
+        return request
+    }
+    
+    static func load(cid: ChannelId, userId: String, context: NSManagedObjectContext) -> ChannelReadDTO? {
+        let request = fetchRequest(for: cid, userId: userId)
+        return try! context.fetch(request).first
+    }
+    
+    static func loadOrCreate(cid: ChannelId, userId: String, context: NSManagedObjectContext) -> ChannelReadDTO {
+        if let existing = Self.load(cid: cid, userId: userId, context: context) {
+            return existing
+        }
+        
+        let new = NSEntityDescription.insertNewObject(forEntityName: Self.entityName, into: context) as! ChannelReadDTO
+        new.channel = ChannelDTO.loadOrCreate(cid: cid, context: context)
+        new.user = UserDTO.loadOrCreate(id: userId, context: context)
+        return new
+    }
+}
+
+// MARK: Saving and loading the data
+
+extension NSManagedObjectContext {
+    func saveChannelRead<ExtraData: ExtraDataTypes>(
+        payload: ChannelReadPayload<ExtraData>,
+        for cid: ChannelId
+    ) throws -> ChannelReadDTO {
+        let dto = ChannelReadDTO.loadOrCreate(cid: cid, userId: payload.user.id, context: self)
+        
+        dto.user = try saveUser(payload: payload.user)
+        
+        dto.lastReadAt = payload.lastReadAt
+        dto.unreadMessageCount = payload.unreadMessagesCount
+        
+        return dto
+    }
+}
+
+extension ChannelReadModel {
+    static func create(fromDTO dto: ChannelReadDTO) -> ChannelReadModel {
+        let user = UserModel<ExtraData.User>.create(fromDTO: dto.user)
+        return .init(lastReadAt: dto.lastReadAt, unreadMessagesCount: dto.unreadMessageCount, user: user)
+    }
+}
```

Comment:
Can you expose this functionality using `asModel()` like we have it for other DTOs? (but Channel - no one fixed it there, yet)

Reply:
I shall fix both!

Reply:
Please don't, it's a huge change. That's why no one done it yet. It's a topic for different PR

Reply:
@b-onc It's related to introducing `ChannelDatabaseSession` and `MemberDatabaseSession` and other changes. It's a tech-debt issue -> CIS-332

Reply:
Fixed in 62b46eb0b0bef360c33b591d1dbceb93446701e2